posts/2024-10-17-Rust-1.82.0.md

```diff
@@ -0,0 +1,319 @@
+
+For several years, the macros `std::ptr::addr_of!` and `std::ptr::addr_of_mut!` have served this purpose. Now the time has come to provide a proper native syntax for this operation: `addr_of!(expr)` becomes `&raw const expr`, and `addr_of_mut!(expr)` becomes `&raw mut expr`. For example:
+
+```rust
+#[repr(packed)]
+struct Packed {
+    not_aligned_field: i32,
+}
+
+fn main() {
+    let p = Packed { not_aligned_field: 1_82 };
+
+    // This would be undefined behavior!
+    // It is rejected by the compiler.
+    //let ptr = &p.not_aligned_field as *const i32;
+
+    // This is the old way of creating a pointer.
+    let ptr = std::ptr::addr_of!(p.not_aligned_field);
+
+    // This is the new way.
+    let ptr = &raw const p.not_aligned_field;
+
+    // Accessing the pointer has not changed.
+    // Note that `val = *ptr` would be undefined behavior because
+    // the pointer is not aligned!
+    let val = unsafe { ptr.read_unaligned() };
+}
+```
+
+The native syntax makes it more clear that the operand expression of these operators is interpreted as a [place expression](https://www.ralfj.de/blog/2024/08/14/places.html). It also avoids the term "address-of" when referring to the action of creating a pointer. A pointer is [more than just an address](https://rust-lang.github.io/rfcs/3559-rust-has-provenance.html), so Rust is moving away from terms like "address-of" that reaffirm a false equivalence of pointers and addresses.
+
+### Safely addressing unsafe `static`s
+
+This code is now allowed:
+
+```rust
+static mut STATIC_MUT: Type = Type::new();
+extern "C" {
+    static EXTERN_STATIC: Type;
+}
+fn main() {
+     let static_mut_ptr = &raw mut STATIC_MUT;
+     let extern_static_ptr = &raw const EXTERN_STATIC;
+}
+```
+
+In an expression context, `STATIC_MUT` and `EXTERN_STATIC` are [place expressions](https://doc.rust-lang.org/reference/expressions.html#place-expressions-and-value-expressions). Previously, the compiler's safety checks were not aware that the raw ref operator did not actually affect the operand's place, treating it as a possible read or write to a pointer. No unsafety is actually present, however, as it just creates a pointer.
+
+Relaxing this may cause problems where some unsafe blocks are now reported as unused if you deny the `unused_unsafe` lint, but they are now only useful on older versions. Annotate these unsafe blocks with `#[allow(unused_unsafe)]` if you wish to support multiple versions of Rust, as in this example diff:
+
+```diff
+ static mut STATIC_MUT: Type = Type::new();
+ fn main() {
++    #[allow(unused_unsafe)]
+     let static_mut_ptr = unsafe { std::ptr::addr_of_mut!(STATIC_MUT) };
```

Comment:
```suggestion
     let static_mut_ptr = unsafe { &raw mut STATIC_MUT };
```

Reply:
This one should *not* change, because if you're writing code that's lint-compatible with older versions, then you can't use `&raw` yet either.